test/e2e/node_feature_discovery.go

```diff
@@ -453,22 +466,96 @@ core:
 
 				By("Verifying node labels from NodeFeatureRules #1 and #2")
 				Expect(waitForNfdNodeLabels(f.ClientSet, expected)).NotTo(HaveOccurred())
+
+				// Add features from NodeFeatureRule #3
+				By("Creating NodeFeatureRules #3")
+				Expect(testutils.CreateNodeFeatureRuleFromFile(nfdClient, "nodefeaturerule-3.yaml")).NotTo(HaveOccurred())
+
+				By("Verifying node taints and annotation from NodeFeatureRules #3")
+				expectedTaints := []corev1.Taint{
+					{
+						Key:    "nfd.node.kubernetes.io/fake-special-node",
+						Value:  "exists",
+						Effect: "NoExecute",
+					},
+					{
+						Key:    "nfd.node.kubernetes.io/fake-dedicated-node",
+						Value:  "true",
+						Effect: "NoExecute",
+					},
+					{
+						Key:    "nfd.node.kubernetes.io/performance-optimized-node",
+						Value:  "true",
+						Effect: "NoExecute",
+					},
+				}
+				expectedAnnotation := map[string]string{
+					"nfd.node.kubernetes.io/taints": "nfd.node.kubernetes.io/fake-special-node=exists:NoExecute,nfd.node.kubernetes.io/fake-dedicated-node=true:NoExecute,nfd.node.kubernetes.io/performance-optimized-node=true:NoExecute"}
+				Expect(waitForNfdNodeTaints(f.ClientSet, expectedTaints)).NotTo(HaveOccurred())
+				Expect(waitForNfdNodeAnnotation(f.ClientSet, expectedAnnotation)).NotTo(HaveOccurred())
+
+				By("Re-applying NodeFeatureRules #3 with updated taints")
+				Expect(testutils.CreateNodeFeatureRuleFromFile(nfdClient, "nodefeaturerule-3-updated.yaml")).NotTo(HaveOccurred())
+				expectedTaintsUpdated := []corev1.Taint{
+					{
+						Key:    "nfd.node.kubernetes.io/fake-special-node",
+						Value:  "exists",
+						Effect: "NoExecute",
+					},
+					{
+						Key:    "nfd.node.kubernetes.io/foo",
+						Value:  "true",
+						Effect: "NoExecute",
+					},
+				}
+				expectedAnnotationUpdated := map[string]string{
+					"nfd.node.kubernetes.io/taints": "nfd.node.kubernetes.io/fake-special-node=exists:NoExecute,nfd.node.kubernetes.io/foo=true:NoExecute"}
+
+				By("Verifying updated node taints and annotation from NodeFeatureRules #3")
+				Expect(waitForNfdNodeTaints(f.ClientSet, expectedTaintsUpdated)).NotTo(HaveOccurred())
+				Expect(waitForNfdNodeAnnotation(f.ClientSet, expectedAnnotationUpdated)).NotTo(HaveOccurred())
 			})
 		})
 	})
 })
 
+// waitForNfdNodeAnnotation waits for node to be annotated as expected.
+func waitForNfdNodeAnnotation(cli clientset.Interface, expectedAnnotation map[string]string) error {
+	poll := func() error {
+		nodeList, err := cli.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
+		if err != nil {
+			return err
+		}
+		for _, node := range nodeList.Items {
+			if diff := cmp.Diff(expectedAnnotation[nfdv1alpha1.NodeTaintsAnnotation], node.Annotations[nfdv1alpha1.NodeTaintsAnnotation]); diff != "" {
+				return fmt.Errorf("node %q annotation does not match expected, diff (expected vs. received): %s", node.Name, diff)
+			}
```

Comment:
As you commented separately, this could be made more generic. I think very easily, i.e. smth like this should work (also renaming the arg `expectedAnnotation` -> `expectedAnnotation` as it would be a map containing all the expected annotations):

```suggestion
			for k, v := range expectedAnnotations {
				if diff := cmp.Diff(v, node.Annotations[k]); diff != "" {
					return fmt.Errorf("node %q annotation does not match expected, diff (expected vs. received): %s", node.Name, diff)
				}
			}
```

WDYT? This should be then readily re-usable when/if we add support for creating node annotations from NodeFeatureRule objects.

Reply:
Good point, thanks. Done.